### PR TITLE
Spark 3.3: Time range query of changelog tables

### PR DESCRIPTION
**Pull request by flyrain**
Per discussion in #6012, I create this PR to support time range query for changelog table. 
cc @aokolnychyi @RussellSpitzer @szehon-ho @rdblue

### CODE REVIEW

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java [hunk 1]
```diff
@@ -32,6 +32,12 @@ private SparkReadOptions() {}
   // End snapshot ID used in incremental scans (inclusive)
   public static final String END_SNAPSHOT_ID = "end-snapshot-id";
 
+  // Start timestamp used in multi-snapshot scans (exclusive)
+  public static final String START_TIMESTAMP = "start-timestamp";
+
+  // End timestamp used in multi-snapshot scans (inclusive)
+  public static final String END_TIMESTAMP = "end-timestamp";
```
Comment (flyrain) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:39]: The other option here is to reuse `as-of-timestamp` as the end timestamp, but I prefer the symmetric options(start-timestamp, end-timestamp). The UX would be better this way, I assume.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 2]
```diff
@@ -308,6 +339,17 @@ public Scan buildChangelogScan() {
     return new SparkChangelogScan(spark, table, scan, readConf, expectedSchema, filterExpressions);
   }
 
+  private Long getStartSnapshotId(Long startTimestamp) {
+    Snapshot oldestSnapshotAfter = SnapshotUtil.oldestAncestorAfter(table, startTimestamp);
+    Preconditions.checkArgument(
```
Comment (flyrain) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:344]: We don't have to throw exception here, another option is to return an empty set of row.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java [hunk 3]
```diff
@@ -32,6 +32,12 @@ private SparkReadOptions() {}
   // End snapshot ID used in incremental scans (inclusive)
   public static final String END_SNAPSHOT_ID = "end-snapshot-id";
 
+  // Start timestamp used in multi-snapshot scans (exclusive)
+  public static final String START_TIMESTAMP = "start-timestamp";
```
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:36]: Looks like Flink has a config called 'start-snapshot-timestamp'.  Its a bit wordy but wondering if we need to synchronize the two (if they are related):  cc @stevenzwu @pvary @hililiwei

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 4]
```diff
@@ -285,6 +286,36 @@ public Scan buildChangelogScan() {
...
+          SparkReadOptions.END_TIMESTAMP);
+    }
+
+    if (startTimestamp != null) {
+      startSnapshotId = getStartSnapshotId(startTimestamp);
+    }
+
+    if (endTimestamp != null) {
+      endSnapshotId = SnapshotUtil.snapshotIdAsOfTime(table, endTimestamp);
+    }
```
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:318]: Do we need to add corresponding checks in build() for not setting these new properties (like we check today for start-snapshot-id)

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (szehon-ho) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:344]: To me, may be slightly more user friendly to return empty set.  But open.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 5]
```diff
@@ -308,6 +339,17 @@ public Scan buildChangelogScan() {
     return new SparkChangelogScan(spark, table, scan, readConf, expectedSchema, filterExpressions);
   }
 
+  private Long getStartSnapshotId(Long startTimestamp) {
+    Snapshot oldestSnapshotAfter = SnapshotUtil.oldestAncestorAfter(table, startTimestamp);
```
Comment (hililiwei) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:343]: Spark processes `startTimestamp` logically differently from Flink. I wonder if we should unify them?

In Flink, it will first find whether there is a snapshot, its time is equal `startTimestamp`. If not, `SnapshotUtil.oldestAncestorAfter` will be used.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java (hunk 3 above)
Comment (hililiwei) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:36]: Yes, it's called "start-snapshot-timestamp" in flink.

https://github.com/apache/iceberg/blob/6d47097151b4df1d3269563a7ebfdb4b6c270a64/flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkReadOptions.java#L48-L49

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java [hunk 6]
```diff
@@ -150,7 +150,7 @@ public static Iterable<Snapshot> ancestorsOf(long snapshotId, Function<Long, Sna
 
   /**
    * Traverses the history of the table's current snapshot and finds the first snapshot committed
-   * after the given time.
+   * after the given time(inclusive).
```
[the remark this answers is not in the document]
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:153]: I'm not sure what Inclusive means in this context? Maybe we need to just move his around to
"finds the oldest snapshot committed at or after the given snapshot"? I dunno
"finds the oldest snapshot that was committed either at or after a given time"?

Or ignore me, I just always find the time docs difficult to understand.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java [hunk 7]
```diff
@@ -137,6 +138,64 @@ public void testOverwrites() {
...
+        "CREATE TABLE %s (id INT, data STRING) "
+            + "USING iceberg "
+            + "PARTITIONED BY (data) "
+            + "TBLPROPERTIES ( "
+            + " '%s' = '%d' "
+            + ")",
+        tableName, FORMAT_VERSION, formatVersion);
+
+    sql("INSERT INTO %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO %s VALUES (2, 'b')", tableName);
```
Comment (RussellSpitzer) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:153]: This section of code is Identical to the test above and the one below,

Probably need some helpers to reduce repeated code in the suite (I also think this helps focus the test code on exactly what is being tested)
createTable()
insertDefaultRows()

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java [hunk 8]
```diff
@@ -137,6 +138,64 @@ public void testOverwrites() {
         changelogRecords(snap2, snap3));
   }
 
+  @Test
+  public void testQueryWithTimeRange() {
```
Comment (RussellSpitzer) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:142]: I would probably split out the invalid cases into another test? Not a big deal though

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java [hunk 9]
```diff
@@ -137,6 +138,64 @@ public void testOverwrites() {
...
+
+    sql("INSERT OVERWRITE %s VALUES (-2, 'b')", tableName);
+
+    table.refresh();
+
+    Snapshot snap3 = table.currentSnapshot();
+
+    assertEquals(
+        "Should have expected changed rows only from snapshot 3",
+        ImmutableList.of(
```
Comment (RussellSpitzer) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:167]: Another way to make this test a bit clearer I would probably make
```
Snapshot1Rows = ImmutableList.of(row(....
Snapshot2Rows = ...
```
then add on a helper
```
rowsWithValue(Snapshot1Rows, snap1.snapshotID)
```


Then this test kind of goes like
```
insertRows(Snapshot1Rows)
insertRows(Snapshot2Rows)
insertRows(Snapshot3Rows)

check(expected = rowsWithValue(Snapshot1Rows, snap1.snapshotID)
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (RussellSpitzer) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:344]: My question would be do we need to differentiate between a snapshot existed yet had no changes and no such snapshot existed.I think i'm good with the exception because it makes it clear you aren't getting results because of the input parameters and not because of the table.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java [hunk 10]
```diff
@@ -137,6 +138,64 @@ public void testOverwrites() {
...
+            row(2, "b", "DELETE", 0, snap3.snapshotId()),
+            row(-2, "b", "INSERT", 0, snap3.snapshotId())),
+        changelogRecords(snap2.timestampMillis() + 1, snap3.timestampMillis()));
```
[the remark this answers is not in the document]
Comment (flyrain) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:170]: Good point, I will use the `waitUntilAfter()` for that.

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java]: Made the change

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java (hunk 3 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:36]: Thanks @szehon-ho and @hililiwei for the input. It'd be nice to consolidate. I kind of prefer the name `start-timestamp` here, so that users don't have to worry about the concept `snapshot`, the query is purely based on the time range.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 4 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:331]: Good point! Added. Although, we could support time range in incremental as well in the future.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 11]
```diff
@@ -285,6 +286,36 @@ public Scan buildChangelogScan() {
 
     Long startSnapshotId = readConf.startSnapshotId();
     Long endSnapshotId = readConf.endSnapshotId();
+    Long startTimestamp = readConf.startTimestamp();
+    Long endTimestamp = readConf.endTimestamp();
+
+    Preconditions.checkArgument(
+        !(startSnapshotId != null && startTimestamp != null),
+        "Cannot set neither %s nor %s for changelogs",
```
[the remark this answers is not in the document]
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java]: Made the change.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java (hunk 7 above)
Comment (flyrain) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java]: Made the change. Thanks for the suggestion.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java (hunk 8 above)
Comment (flyrain) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:123]: Done.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java (hunk 9 above)
Comment (flyrain) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestChangelogTable.java:137]: Tried it. We have to change the `_change_ordinal` as well. Need a bit more logic to make it work, it may not be worth considering there only about 3 test cases using that. WDYT?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 5 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:356]: Is it inclusive for the `startTimestamp` in Flink?
>Spark processes startTimestamp logically differently from Flink. I wonder if we should unify them?

Is it necessary to do that since `oldestAncestorAfter(startTimestamp)` will return the snapshot whose timestamp equals to `startTimestamp`. 
>In Flink, it will first find whether there is a snapshot, its time is equal startTimestamp. If not, SnapshotUtil.oldestAncestorAfter will be used.

**Comment by flyrain** [comment]
Hi @szehon-ho, @RussellSpitzer, @hililiwei thanks for the review! Resolved your comments. Take another look?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:357]: > do we need to differentiate between a snapshot existed yet had no changes and no such snapshot existed.

I will consider these two use cases.
1. Query with snapshot id. We should differentiate it, throwing an exception makes sense here.
2. Query only with time range. We may assume user doesn't have to know the concept of `snapshot`. We may not differentiate it, returning an empty set makes more sense.

Subtle difference though. I am fine with either option.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 5 above)
Comment (hililiwei) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:356]: > Is it necessary to do that since oldestAncestorAfter(startTimestamp) will return the snapshot whose timestamp equals to startTimestamp.

I have re-checked and you are right. Then,  I raised https://github.com/apache/iceberg/pull/6401 to try to unify them.

Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:356]: Yeah, I was confused by the name at the beginning as well.

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java [hunk 12]
```diff
@@ -149,8 +149,7 @@ public static Iterable<Snapshot> ancestorsOf(long snapshotId, Function<Long, Sna
   }
 
   /**
-   * Traverses the history of the table's current snapshot and finds the first snapshot committed
-   * after the given time.
+   * Finds the oldest snapshot that was committed either at or after a given time.
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:152]: Nit: is it necessary to change whole sentence?  How about just replacing 'after' to 'at or after'?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 13]
```diff
@@ -213,13 +214,26 @@ public Scan build() {
...
+
     if (startSnapshotId != null) {
       return buildIncrementalAppendScan(startSnapshotId, endSnapshotId);
     } else {
       return buildBatchScan(snapshotId, asOfTimestamp, branch, tag);
     }
   }
 
+  private void checkInvalidOptions() {
+    Long startTimestamp = readConf.startTimestamp();
```
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:227]: Nit: maybe just inline this if its only used in one place (unless I miss something)?  Doesnt make sense to make a helper method called 'checkInvalidOption' but only add one specific check.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (szehon-ho) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:357]: Having an exception to distinguish is a fair point.  Initially I was thinking it would be a pain for user to have to deal with exception if they wanted to do all changes from a random time, but they can find the earliest snapshot and make sure the timestamp is appropriate to avoid exception.  So still no real preference, ok either way.

**Review by szehon-ho — COMMENTED** [review]
This looks good to me, just some minor naming nits.

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java (hunk 12 above)
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:152]: The first part of the comment is confusing, it should be at least something like this. 
```
Traverses the table's snapshot history
```
Moreover, I think it is not necessary to mention that. WDYT?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 13 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:227]: I thought it'd be more readable. No? I'm OK with either.

Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:217]: Inlined in the new commit.

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java (hunk 12 above)
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java]: Rephrased a bit. Please take a look. Any suggestion is welcome.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java (hunk 2 above)
Comment (RussellSpitzer) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:353]: I think you are correct @flyrain , I only really would want the error for 1.

**Review by RussellSpitzer — APPROVED** [review]
Looking good to me, Once everyone else who is reviewing has signed off I think we are good to merge

**Comment by flyrain** [comment]
Thanks @RussellSpitzer. Hi @szehon-ho @hililiwei , please let me know if you have any comments. Thanks1

**Comment by flyrain** [comment]
Thanks @szehon-ho for the review. I believe you are talking about the case 2 in https://github.com/apache/iceberg/pull/6350#discussion_r1044906141. I did try to return an empty set, but it seems a bit involved to make the change. Solutions would be
1. Add a new interface to interface `IncrementalScan` to return an empty set.
2. Reuse `IncrementalScan::fromSnapshotExclusive` by passing the latest/current snapshot id. It is a workaround without interface change. I tried this, which needs to touch a bunch of code.

Let me know if there are easier way for that. Otherwise, I will make the change in another PR.

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java [hunk 14]
```diff
@@ -149,8 +149,7 @@ public static Iterable<Snapshot> ancestorsOf(long snapshotId, Function<Long, Sna
   }
 
   /**
-   * Traverses the history of the table's current snapshot and finds the first snapshot committed
-   * after the given time.
+   * Finds the oldest snapshot of a table that was committed either at or after a given time.
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:152]: @flyrain, this change is incorrect. This is not the oldest known snapshot, it is the oldest known snapshot in the current snapshot's history. Can you fix it?

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:152]: Sure, will do.

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:152]: Filed https://github.com/apache/iceberg/pull/6479
